optuna/samplers/_motpe.py

```diff
@@ -0,0 +1,627 @@
+                    - hv_selected
+                )
+            selected_vecs += [selected_vec]
+            selected_indices += [selected_index]
+            hv_selected = self._compute_hypervolume(np.asarray(selected_vecs), reference_point)
+
+        return np.asarray(selected_indices, dtype=int)
+
+    def _calculate_weights_below(
+        self,
+        lvals: np.ndarray,
+        indices_below: np.ndarray,
+    ) -> np.ndarray:
+        # Calculate weights based on hypervolume contributions.
+        n_below = len(indices_below)
+        if n_below == 0:
+            return np.asarray([])
+        elif n_below == 1:
+            return np.asarray([1.0])
+        else:
+            lvals_below = lvals[indices_below].tolist()
+            worst_point = np.max(lvals_below, axis=0)
+            reference_point = np.maximum(1.1 * worst_point, 0.9 * worst_point)
+            reference_point[reference_point == 0] = EPS
+            hv = self._compute_hypervolume(np.asarray(lvals_below), reference_point)
+            contributions = np.asarray(
+                [
+                    hv
+                    - self._compute_hypervolume(
+                        np.asarray(lvals_below[:i] + lvals_below[i + 1 :]), reference_point
+                    )
+                    for i in range(len(lvals))
+                ]
+            )
+            weights_below = np.clip(contributions / np.max(contributions), 0, 1)
+            return weights_below
+
+
+def _calculate_nondomination_rank(loss_vals: np.ndarray) -> np.ndarray:
+    vecs = loss_vals.copy()
+
+    # Normalize values
+    lb = vecs.min(axis=0, keepdims=True)
+    ub = vecs.max(axis=0, keepdims=True)
+    vecs = (vecs - lb) / (ub - lb)
+
+    ranks = np.zeros(len(vecs))
+    num_unranked = len(vecs)
+    rank = 0
+    while num_unranked > 0:
+        extended = np.tile(vecs, (vecs.shape[0], 1, 1))
+        counts = np.sum(
+            np.logical_and(
+                np.all(extended <= np.swapaxes(extended, 0, 1), axis=2),
+                np.any(extended < np.swapaxes(extended, 0, 1), axis=2),
+            ),
+            axis=1,
+        )
+        vecs[counts == 0] = 1.1  # mark as ranked
+        ranks[counts == 0] = rank
+        rank += 1
+        num_unranked -= np.sum(counts == 0)
+    return ranks
+
+
+def _get_observation_pairs(
+    study: optuna.study.Study,
+    param_name: str,
+) -> Tuple[List[Optional[float]], List[List[float]]]:
+    """Get observation pairs from the study.
+
+    This function collects observation pairs from the complete trials of the study.
+    Pruning is currently not supported.
+    The values for trials that don't contain the parameter named ``param_name`` are set to None.
+
+    Objective values are negated if their directions are maximization and all objectives are
+    treated as minimization in the MOTPE algorithm.
+    """
+
+    trials = study._storage.get_all_trials(study._study_id, deepcopy=False)
```

Comment:
Can be a follow up but you can optimize/simplify this and the following code using the following.
```suggestion
    trials = study._storage.get_all_trials(study._study_id, deepcopy=False, states=(optuna.trial.TrialState.COMPLETE,))
```